### ISSUE TITLE
Autograd.grad not making higher order graph for scaling functions

### ISSUE THREAD

**Issue by KeAWang** [pytorch/pytorch#36463]
## 🐛 Bug

`torch.autograd.grad` seems to not create graph if the function being differentiated has a second derivative of 0. This causes unexpected issues when we try to take a higher-order gradient.

The following example will make this more clear

## To Reproduce
We would like to do something like
```python
import torch

def f(x):
    return (2 * x).sum()

x = torch.randn(5).requires_grad_()
y = f(x)
dydx = torch.autograd.grad(y, x, create_graph=True)[0]
d2ydx2 = torch.autograd.grad(dydx, x)[0]
```
This fails because `dydx` doesn't have a `grad_fn` even though `create_graph=True`. This seems to be because `dydx` is a constant, as expected. If we replace `f` with `lambda x: x.pow(2).sum()` then we retain the `grad_fn`. 

However, in general, if we may not know what `f` is analytically so autograd should keep a `grad_fn` on `dydx` so that we can compute higher order gradients without having to check if `dydx` retained its `grad_fn`.

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: EndeavourOS Linux
GCC version: (GCC) 8.4.0
CMake version: version 3.17.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 440.82
cuDNN version: /usr/lib/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] pytorch-lightning==0.7.4.dev0
[pip] torch==1.4.0
[pip] torchcontrib==0.0.2
[pip] torchdiffeq==0.0.1
[pip] torchvision==0.5.0
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.0                      166    conda-forge
[conda] numpy                     1.18.1           py37h8960a57_1    conda-forge
[conda] pytorch                   1.4.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-lightning         0.7.4.dev0                dev_0    <develop>
[conda] torchcontrib              0.0.2                    pypi_0    pypi
[conda] torchdiffeq               0.0.1                    pypi_0    pypi
[conda] torchvision               0.5.0                py37_cu101    pytorch

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Hi,

Sorry for the late reply.
This is the expected behavior actually. Unfortunately creating a full graph for these kind of things would lead to significant performance regression (as the graph is currently optimized away).

In pytorch, a gradient with value 0 is represented by a Tensor full of 0s but an input independent of an output can be represented by:
- A Tensor containing 0 (after indexing ops for example)
- None (returned by autograd.grad(.., allow_unused=True))
- A disconnected graph

So if you get None or a disconnected graph, you can be sure that the higher order derivatives will all be `0` as well (because they're independent).
Does that answer your question?

**Comment by KeAWang** [comment]
I see. Thank you!
